[MNG-7914] Provide a single entry point for configuration

gnodet opened the pull request.

michael-o reviewed (changes requested):
There is a conceptual problem here. There are only Maven user properties and Java system properties, but not Maven system properties. We are trying to get rid of them. That should clear in a new solution. I wouldn't even try to make system properties right. @slawekjaranowski

gnodet commented:
> There is a conceptual problem here. There are only Maven user properties and Java system properties, but not Maven system properties. We are trying to get rid of them. That should clear in a new solution. I wouldn't even try to make system properties right. @slawekjaranowski

Yes, it makes sense.  I've removed the system properties support to only keep `maven.properties` which are user properties.

michael-o commented:
There is one conceptual problem I have here: How does this relate to `maven.config`  doesn't it cause confusion? Add even more complexity?

michael-o commented:
> > There is one conceptual problem I have here: How does this relate to `maven.config` doesn't it cause confusion? Add even more complexity?
> 
> It's imho, a much better way to provide configuration properties. `maven.config` can be used to provide _command line arguments_, but it has a few problems: it does not have the usual multi-level that we have for settings or toolchains (global, user, project). In addition, the syntax is not user friendly, does not support inclusions of other files and does not support substitutions. Imho, we should keep `maven.config` for actual CLI arguments and users should use `maven.properties` to define properties such as https://maven.apache.org/resolver/configuration.html (we should have a similar one for maven).

I see, then we also should educate users when to use what.

michael-o commented on apache-maven/src/assembly/maven/conf/maven.properties:
```diff
@@ -0,0 +1,10 @@
+${includes} =
+
+# Optionally load global user properties and project user properties
+${optionals} = ${user.home}/.m2/maven.properties \
```
Complete side note: We should long term reconsider getting rid of `.m2` altogether and move to a more generic terlm like `.maven`.

gnodet replied:
That makes sense.  At least, in the short term, we can make it configurable and try to change the hardcoded references in the code to something more flexible.  I'll work on a follow-up PR, I'm already working on generating documentation for maven properties.

gnodet commented:
> > > There is one conceptual problem I have here: How does this relate to `maven.config` doesn't it cause confusion? Add even more complexity?
> > 
> > 
> > It's imho, a much better way to provide configuration properties. `maven.config` can be used to provide _command line arguments_, but it has a few problems: it does not have the usual multi-level that we have for settings or toolchains (global, user, project). In addition, the syntax is not user friendly, does not support inclusions of other files and does not support substitutions. Imho, we should keep `maven.config` for actual CLI arguments and users should use `maven.properties` to define properties such as https://maven.apache.org/resolver/configuration.html (we should have a similar one for maven).
> 
> I see, then we also should educate users when to use what.

Agreed.  I suppose enhancing https://maven.apache.org/configure.html would be a first step.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/props/InterpolationHelper.java:
```diff
@@ -0,0 +1,364 @@
+            Map<String, String> cycleMap,
+            Map<String, String> configProps,
+            SubstitutionCallback callback)
+            throws IllegalArgumentException {
```
The throws in the signature is redundant for runtime exceptions and we have good Javadoc here.

gnodet replied:
I'll fix that.  Fwiw, those three classes come from Karaf/Felix.  They even provide a bit more, as they can even be used to programmatically modify properties file while keeping layout and comments (though I don't think we'll need that in the short term), or even typed properties (not included here).

gnodet replied:
Fixed

michael-o commented:
I will get back to this tomorrow. This is a crucial change which needs careful design.

michael-o commented on apache-maven/src/assembly/maven/conf/maven.properties:
```diff
@@ -0,0 +1,41 @@
+
+# Space separated list of files to load.
+# If the first character is a question mark, the load will silently fail if the file does not exist.
+${includes} = ?${maven.user.conf}/maven.properties, \
```
There is a logical question here to what extend should the question mark apply?  Until the last "}"? From a logical PoV it should not include `/maven.properties.

michael-o replied:
I thinking whether there are already used syntaxes people might know already.

michael-o commented on maven-api-impl/src/main/java/org/apache/maven/internal/impl/DefaultSettingsBuilder.java:
```diff
@@ -66,8 +66,8 @@ public class DefaultSettingsBuilder implements SettingsBuilder {
 
-        Source globalSource = request.getGlobalSettingsSource().orElse(null);
-        Settings global = readSettings(globalSource, false, request, problems);
+        Source installSource = request.getInstallationSettingsSource().orElse(null);
```
`installationSource`

michael-o commented on maven-api-impl/src/main/java/org/apache/maven/internal/impl/DefaultToolchainsBuilder.java:
```diff
@@ -57,13 +57,13 @@ public class DefaultToolchainsBuilder implements ToolchainsBuilder {
 
-        Source globalSource = request.getGlobalToolchainsSource().orElse(null);
-        PersistedToolchains global = readToolchains(globalSource, request, problems);
+        Source installSource = request.getInstallationToolchainsSource().orElse(null);
```
Same here

michael-o commented on maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/MavenBuildTimestamp.java:
```diff
@@ -70,7 +67,6 @@ public MavenBuildTimestamp(Instant time, String timestampFormat) {
         }
         SimpleDateFormat dateFormat = new SimpleDateFormat(timestampFormat);
         dateFormat.setCalendar(new GregorianCalendar());
-        dateFormat.setTimeZone(DEFAULT_BUILD_TIME_ZONE);
```
Attention, this is a semantic change. We should retain `Etc/UTC` and maybe make it configurable in the future.

gnodet replied:
Why is that a change ? We added the timezone offset in the description instead of using `Z`, so we need to use the current time zone, else, the formatted time will be wrong I think.

michael-o replied:
Without setting the timezone it will use the local timezone. It won't be reproducible for those crossing time zones. If you set the timezone `XXX` will produce stable output, with `Etc/UTC` always `Z`. Then if we say that the timezone can be configured `XXX` will produce proper output, `'Z'` is just a static literal.

michael-o replied:
https://reproducible-builds.org/docs/timezones/

Try it out in a `main()` and you will see the difference.

michael-o commented on maven-core/src/main/java/org/apache/maven/project/DefaultProjectBuilder.java:
```diff
@@ -358,10 +359,7 @@ public void close() {
-                if (str == null) {
-                    str = request.getSystemProperties().getProperty(BUILDER_PARALLELISM);
-                }
+                String str = request.getUserProperties().getProperty(Constants.MAVEN_PROJECT_BUILDER_PARALLELISM);
```
Very good, less dependence on system properties

michael-o commented on apache-maven/src/assembly/maven/conf/maven.properties:
```diff
@@ -0,0 +1,40 @@
+# Comma-separated list of files to include.
+# Each item may be enclosed in quotes to gracefully include spaces. Items are trimmed before being loaded.
+# If the first character of an item is a question mark, the load will silently fail if the file does not exist.
+${includes} = "?${maven.user.conf}/maven.properties", \
```
Since you are quoting now the value, wouldn't it be more natural to put the question mark at the end? Unless you want to use "¿" as well. Maybe qoutes should be mandatory. Worth looking into Tomcat code for ths. I do not remember the motiviation.

gnodet replied:
I'm not sure quotes are mandatory, but they certainly avoid any problem with spaces inside path.  
About the location of the '?', I fear that putting it at end will be less prominent.
It's also the same syntax than for optional profiles, see https://maven.apache.org/guides/introduction/introduction-to-profiles.html#explicit-profile-activation

michael-o replied:
I see your point. Shouldn't then the question mark be outside of the quotation marks to make the entire space-containing value optional?

gnodet replied:
Yes, makes sense.

gnodet replied:
I've added support for moving the '?' before the quotes (both before and after are supported), the default config file moves them before.

michael-o commented on maven-api-impl/src/main/java/org/apache/maven/internal/impl/model/MavenBuildTimestamp.java:
```diff
@@ -26,18 +26,18 @@
-    public static final String DEFAULT_BUILD_TIMESTAMP_FORMAT = "yyyy-MM-dd'T'HH:mm:ss'Z'";
-
-    public static final String BUILD_TIMESTAMP_FORMAT_PROPERTY = "maven.build.timestamp.format";
+    public static final String DEFAULT_BUILD_TIMESTAMP_FORMAT = "yyyy-MM-dd'T'HH:mm:ssXXX";
```
Why not use `Constants.MAVEN_BUILD_TIMESTAMP_FORMAT` instead of a literal again?

gnodet replied:
`Constants.MAVEN_BUILD_TIMESTAMP_FORMAT` defines the property name.  The default value is available on the `@Config` annotation, but at this point, it is for documentation only.  

Another way would be to move the default values as constants, i.e. move DEFAULT_BUILD_TIMESTAMP_FORMAT to `Constants` and link both there.

michael-o replied:
Yes, that seems to make sense. No need to define it twice, yes?

gnodet replied:
I've had a look, unfortunately, most of those properties don't really use a simple value.
A bunch of them use property substitution such ac `${maven.home}/conf`, or even non computable values, such as `cores/2 + 1`.  It would only work for a few of them, so I'd rather keep it that way.

michael-o replied:
That's fine.

slawekjaranowski commented on maven-embedder/src/main/java/org/apache/maven/cli/props/MavenPropertiesLoader.java:
```diff
@@ -0,0 +1,176 @@
+                            mandatory = false;
+                            location = location.substring(1);
+                        }
+                        Path path = configProp.resolve(location);
```
proposition:
```
                        Path path = configProp.getParent().resolve(location);
```

`path` is a path for current properties which contains includes so we should resolve against directory contains properties - not a property file.

it will allow relative path in properties like:
```
${includes}=default.properties
```

gnodet replied:
That was the intent, it was clearly missing a unit test.

slawekjaranowski commented on maven-embedder/src/main/java/org/apache/maven/cli/props/InterpolationHelper.java:
```diff
@@ -0,0 +1,353 @@
+    private static final String DELIM_START = "${";
+    private static final String DELIM_STOP = "}";
+    private static final String MARKER = "$__";
+    private static final String ENV_PREFIX = "env:";
```
Why not: `env.` - it is more natural as we can use `${env.NAME}`

michael-o replied:
Good catch. It should be consistent with our current approach.

gnodet replied:
Actually, I'll remove that part, as we do pass the system+"env." properties for substitution already.

slawekjaranowski commented:
my test in project `.mvn/maven.properties` with content:
```
${includes}=default.properties,?env-${env:envName}.properties
```

looks like `${env...}` is not interpolated in project properties

executed:
```
envName=test mvn ...
```

It will be great to support user properties in such case to have a possibility:

```
${includes}=default.properties,?env-${envName}.properties
```
and 

```
mvn -DenvName=test ...
```

slawekjaranowski commented on maven-embedder/src/main/java/org/apache/maven/cli/props/MavenPropertiesLoader.java:
```diff
@@ -0,0 +1,176 @@
+        MavenProperties configProps = new MavenProperties(null, false);
+        try {
+            configProps.load(path);
+        } catch (NoSuchFileException ex) {
+            if (failIfNotFound) {
+                throw ex;
+            }
```
Here we can have other IOException, eg I have:

```
Error loading config properties from /Users/..../repos/test/test/.mvn/maven.properties/default.properties
Main: java.nio.file.FileSystemException: /Users/..../repos/test/test/.mvn/maven.properties/default.properties: Not a directory
```

Maybe we should check if file exist?

gnodet replied:
Fixed

gnodet commented:
> my test in project `.mvn/maven.properties` with content:
> 
> ```
> ${includes}=default.properties,?env-${env:envName}.properties
> ```
> 
> looks like `${env...}` is not interpolated in project properties
> 
> executed:
> 
> ```
> envName=test mvn ...
> ```
> 
> It will be great to support user properties in such case to have a possibility:
> 
> ```
> ${includes}=default.properties,?env-${envName}.properties
> ```
> 
> and
> 
> ```
> mvn -DenvName=test ...
> ```

I've added a unit test showing this now works.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -510,7 +503,8 @@ void logging(CliRequest cliRequest) {
         // LOG COLOR
-        String styleColor = cliRequest.getUserProperties().getProperty(STYLE_COLOR_PROPERTY, "auto");
+        String styleColor = cliRequest.getUserProperties().getProperty("style.color", "auto");
+        styleColor = cliRequest.getUserProperties().getProperty(Constants.MAVEN_STYLE_COLOR_PROPERTY, styleColor);
```
Is that on purpose?

gnodet replied:
Yes, I wanted all public properties in `Constants` to start with `maven.`.  That one was an exception, so it's now used if the new correct one isn't set.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -874,9 +873,9 @@ private static <T> List<T> reverse(List<T> list) {
+        String extClassPath = cliRequest.userProperties.getProperty(Constants.MAVEN_EXT_CLASS_PATH);
         if (extClassPath == null) {
-            extClassPath = cliRequest.systemProperties.getProperty(EXT_CLASS_PATH);
+            extClassPath = cliRequest.systemProperties.getProperty(Constants.MAVEN_EXT_CLASS_PATH);
```
Can we add a note whether this should still be system property? If we are already in Java then we obviously don't need a system property at all, only user one.

See: https://github.com/apache/maven/pull/1595/files#diff-b8b814f5b6b3855ae79dc45a0266b94871193dcef42803c316e07409e94af35cR1404

gnodet replied:
I think we can keep a user properties only.
What about adding a warning if the user uses this system property ? We could direct him to use a user property instead ?

gnodet replied:
@michael-o I've added a warning if a system property is used.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1204,40 +1203,57 @@ void toolchains(CliRequest cliRequest) throws Exception {
-                "Reading global toolchains from '{}'",
-                getLocation(toolchainsRequest.getGlobalToolchainsSource(), globalToolchainsFile));
+                "Reading installation toolchains from '{}'",
+                getLocation(toolchainsRequest.getGlobalToolchainsSource(), installationToolchainsFile));
```
Do you intend to deprecate old getters/setters as well (later)?

gnodet replied:
I haven't touched the Maven 3 API at this point.   When we consider the Maven 4 api stable and complete enough, I think we'll have to deprecate the whole Maven 3 API, so I would think this will be part of that change.

michael-o replied:
Makes sense!

slawekjaranowski commented on maven-embedder/src/main/java/org/apache/maven/cli/props/MavenPropertiesLoader.java:
```diff
@@ -0,0 +1,161 @@
+        MavenProperties sp = new MavenProperties(false);
+        try {
+            sp.load(path);
+        } catch (NoSuchFileException ex) {
```
Should we check also other IOException or path exist also here?

gnodet replied:
I've added a check for the existence of the file and let the exception bubble up. The code is nicer.

slawekjaranowski commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1624,20 +1635,65 @@ static void populateProperties(
+        // ----------------------------------------------------------------------
+        // Load config files
+        // ----------------------------------------------------------------------
+        Function<String, String> callback =
+                or(paths::getProperty, prefix("cli.", commandLine::getOptionValue), systemProperties::getProperty);
```
Please also add: `userProperties::getProperty`, so we can use: `mvn ... -Dname=xxx`

michael-o replied:
Good catch!

gnodet replied:
This should not be necessary.   The command line options are added to the `userProperties` a few line earlier:
https://github.com/apache/maven/blob/c4fa7a099a7f338e5c78b93fb476f9d451349dda/maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java#L1644-L1646
So all those user-provided properties will be interpolated the same way than those from the properties file (and thus be available for being themselves values for interpolation).

slawekjaranowski replied:
it is necessary

slawekjaranowski replied:
Properties for testing: 
```
${includes}=default.properties,env-${envName}.properties
```

command:
```
mvn -v -DenvName=test
```

fail with:

```
java.nio.file.NoSuchFileException: ..../repos/test/test/.mvn/env-${envName}.properties
```

gnodet commented:
@michael-o @slawekjaranowski @cstamas any more comments on that one ?

michael-o commented:
> @michael-o @slawekjaranowski @cstamas any more comments on that one ?

Will give it a spin again today

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -874,9 +873,17 @@ private static <T> List<T> reverse(List<T> list) {
+                                + "The property can be passed as a maven argument or in the maven project configuration file,"
+                                + "usually located at {}.",
+                        Constants.MAVEN_EXT_CLASS_PATH,
+                        "[rootDirectory]/.mvn/maven.properties");
```
This message should use prop syntax: `"${rootDirectory}/.mvn/maven.properties"}`, no?

gnodet replied:
Makes sense.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1204,40 +1211,57 @@ void toolchains(CliRequest cliRequest) throws Exception {
+            if (installationToolchainsFileStr != null) {
+                installationToolchainsFile = new File(installationToolchainsFileStr);
+                installationToolchainsFile = resolveFile(installationToolchainsFile, cliRequest.workingDirectory);
+            }
```
This entire block makes me think: WE fail here if the toolchains file cannot be loaded, but we do not fail if core extensions file is provided, but not valid? FNFE and frieds...

michael-o replied:
Or is it `{maven.rootDirectory}`/`{project.rootDirectory}`, anyway it should be consistent with the interpolation style.

gnodet replied:
Then it should be `${session.rootDirectory}`

michael-o replied:
> Then it should be `${session.rootDirectory}`

That'd be perfect.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -874,9 +873,17 @@ private static <T> List<T> reverse(List<T> list) {
+                slf4jLogger.warn(
+                        "The property {} has been set using a JVM system property which is deprecated. "
+                                + "The property can be passed as a maven argument or in the maven project configuration file,"
+                                + "usually located at {}.",
```
Args should be surrounded by single quotes.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1204,40 +1211,57 @@ void toolchains(CliRequest cliRequest) throws Exception {
         }
 
-        File globalToolchainsFile;
+        File installationToolchainsFile = null;
```
Why not use `Path` here throughout?

gnodet replied:
No reason, but I'd like to not expand the scope of this PR too much.  I'd rather raise another PR for doing a full pass on `MavenCli` to switch to nio2 Path.

michael-o replied:
Agreed.

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1380,14 +1404,14 @@ private static boolean isRunningOnCI(Properties systemProperties) {
         // TODO Investigate why this can also be a Java system property and not just a Maven user property like
         // other properties
-        return request.getSystemProperties().getProperty(MavenCli.LOCAL_REPO_PROPERTY);
+        return request.getSystemProperties().getProperty(Constants.MAVEN_REPO_LOCAL);
```
@gnodet, should this also issue a warning like you did with ext classpath being a system property?

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1624,20 +1643,65 @@ static void populateProperties(
+        userProperties.stringPropertyNames().stream()
+                .filter(k -> !sys.contains(k))
+                .forEach(k -> System.setProperty(k, userProperties.getProperty(k)));
+    }
```
We must seriously reconsider this promotion in the future before GA.

slawekjaranowski replied:
The best we should not change a System property at all

gnodet replied:
I've raised https://github.com/apache/maven/pull/1661 to assess if there are any failing ITs as a first step.

michael-o replied:
Invoker and friends are our worst enemies here.

michael-o reviewed (changes requested):
What is lacking completely is documentation how feature rich this approach is with inclusiion and optional files w/o reading source code.

slawekjaranowski reviewed (changes requested):
evaluate properties from cmd

gnodet commented:
> evaluate properties from cmd

Could you expand a bit more ?

slawekjaranowski commented:
> > evaluate properties from cmd
> 
> Could you expand a bit more ?

more in comments: https://github.com/apache/maven/pull/1595#pullrequestreview-2172735294

here only as reason why I marek PR as require change

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -878,11 +878,10 @@ private List<File> parseExtClasspath(CliRequest cliRequest) {
                 slf4jLogger.warn(
-                        "The property {} has been set using a JVM system property which is deprecated. "
+                        "The property '{}' has been set using a JVM system property which is deprecated. "
                                 + "The property can be passed as a maven argument or in the maven project configuration file,"
```
Maven should be uppercase since it is a name

michael-o commented on maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java:
```diff
@@ -1405,13 +1404,17 @@ private static boolean isRunningOnCI(Properties systemProperties) {
+            if (userDefinedLocalRepo != null) {
+                slf4jLogger.warn(
+                        "The property '{}' has been set using a JVM system property which is deprecated. "
+                                + "The property can be passed as a maven argument or in the maven project configuration file,"
```
Maven should be uppercase since it is a name

gnodet commented:
> > > evaluate properties from cmd
> > 
> > 
> > Could you expand a bit more ?
> 
> more in comments: [#1595 (review)](https://github.com/apache/maven/pull/1595#pullrequestreview-2172735294)
> 
> [#1595 (comment)](https://github.com/apache/maven/pull/1595#discussion_r1720812179)
> 
> here only as reason why I marek PR as require change

Thx.  I modified the UT and fixed the use case.

slawekjaranowski reviewed (approved):
Looks ok for me.

We need to review documentation for it at all, generated and manually written.

eolivelli reviewed (approved):
Great
